### PR TITLE
Add Question 189: Compute Direct Preference Optimization Loss

### PR DESCRIPTION
**Pull request by zhenhuan-yang**
## Summary
This PR adds a new medium-difficulty Deep Learning question on computing Direct Preference Optimization (DPO) loss for language model alignment.

## Question Details
- **ID**: 189
- **Title**: Compute Direct Preference Optimization Loss
- **Difficulty**: Medium
- **Category**: Deep Learning

## Implementation
- ✅ Complete solution with proper numerical stability using `np.log1p`
- ✅ Comprehensive educational content covering DPO theory and Bradley-Terry model
- ✅ Mathematical formulation with LaTeX
- ✅ 4 diverse test cases with varying parameters
- ✅ Example with detailed reasoning

## Validation
- ✅ Build successful
- ✅ Schema validation passed
- ✅ All test cases pass

## Educational Value
Covers an important modern technique for LLM alignment that's simpler and more stable than traditional RLHF, making it highly relevant for current ML practitioners.

### CODE REVIEW

File: questions/189_compute-direct-preference-loss/solution.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    # Compute log ratios for chosen and rejected responses
+    policy_log_ratios = policy_chosen_logps - policy_rejected_logps
+    reference_log_ratios = reference_chosen_logps - reference_rejected_logps
+
+    # Compute the logits for the Bradley-Terry model
+    logits = beta * (policy_log_ratios - reference_log_ratios)
+
+    # Compute loss using log-sigmoid for numerical stability
+    # Loss = -log(sigmoid(logits)) = log(1 + exp(-logits))
+    losses = np.log1p(np.exp(-logits))
```
Comment (moe18) [inline questions/189_compute-direct-preference-loss/solution.py:28]: This overflows to inf when logits is large and negative
```losses = np.logaddexp(0, -logits)```
this might be more stable

Comment (moe18) [reply questions/189_compute-direct-preference-loss/solution.py:28]: I would also add a test case to exploit this
